nexus/types/src/deployment/blueprint_display.rs

```diff
@@ -0,0 +1,331 @@
+
+    pub const UNCHANGED_PARENS: &str = "(unchanged)";
+    pub const NONE_PARENS: &str = "(none)";
+    pub const NOT_PRESENT_IN_COLLECTION_PARENS: &str =
+        "(not present in collection)";
+}
+use constants::*;
+
+/// The state of a sled or resource (e.g. zone or physical disk) in this
+/// blueprint, with regards to the parent blueprint
+#[derive(Debug, Clone, Copy)]
+pub enum BpDiffState {
+    Unchanged,
+    Removed,
+    Modified,
+    Added,
+}
+
+impl BpDiffState {
+    pub fn prefix(&self) -> char {
+        match self {
+            BpDiffState::Unchanged => UNCHANGED_PREFIX,
+            BpDiffState::Removed => REMOVED_PREFIX,
+            BpDiffState::Modified => MODIFIED_PREFIX,
+            BpDiffState::Added => ADDED_PREFIX,
+        }
+    }
+}
+
+impl fmt::Display for BpDiffState {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s = match self {
+            BpDiffState::Unchanged => "UNCHANGED",
+            BpDiffState::Removed => "REMOVED",
+            BpDiffState::Modified => "MODIFIED",
+            BpDiffState::Added => "ADDED",
+        };
+        write!(f, "{s}")
+    }
+}
+
+/// A wrapper aound generation numbers for blueprints or blueprint diffs
+#[derive(Debug, Clone, Copy)]
+pub enum BpGeneration {
+    // A value in a single blueprint
+    Value(Generation),
+
+    // A diff between two blueprints
+    Diff { before: Option<Generation>, after: Option<Generation> },
+}
+
+impl fmt::Display for BpGeneration {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            BpGeneration::Value(generation) => {
+                write!(f, "at generation {generation}")
+            }
+            BpGeneration::Diff { before: None, after: Some(after) } => {
+                write!(f, "at generation {after}")
+            }
+            BpGeneration::Diff { before: Some(before), after: None } => {
+                write!(f, "from generation {before}")
+            }
+            BpGeneration::Diff { before: Some(before), after: Some(after) } => {
+                if before == after {
+                    write!(f, "at generation {after}")
+                } else {
+                    write!(f, "generation {before} -> {after}")
+                }
+            }
+            BpGeneration::Diff { before: None, after: None } => {
+                write!(f, "Error: unknown generation")
+            }
+        }
+    }
+}
+
+/// A row in a [`BpSledSubtable`]
+pub struct BpSledSubtableRow {
+    state: BpDiffState,
+    columns: Vec<String>,
+}
+
+impl BpSledSubtableRow {
+    pub fn new(state: BpDiffState, columns: Vec<String>) -> Self {
+        BpSledSubtableRow { state, columns }
+    }
+}
+
+/// Metadata about all instances of specific type of [`BpSledSubtable`],
+/// such as omicron zones or physical disks.
+pub trait BpSledSubtableSchema {
+    fn table_name(&self) -> &'static str;
+    fn column_names(&self) -> Vec<&'static str>;
```

Comment:
This requires a new `Vec` every time it's called. Probably fine for diffs, but could it be `&[&'static str]` instead?

Reply:
Fixed.